metals/src/main/scala/scala/meta/internal/parsing/FoldingRangeExtractor.scala

```diff
@@ -148,108 +142,211 @@ final class FoldingRangeExtractor(
-          } yield (pos, true)
+          matchBlockStartAndFoldingAdjustment[Token.KwMatch](
+            term.expr.trailingTokens
+          ) match {
+            case Some(foldStartIndexAndAdjust) =>
+              Some(
+                range(
+                  tree.pos.input,
+                  foldStartIndexAndAdjust._1,
+                  term.pos.end,
+                ),
+                foldStartIndexAndAdjust._2,
+              )
+            case None => None
+          }
 
         case c: Case =>
           val startingPoint = c.cond.getOrElse(c.pat)
-          val bodyEnd = {
-            c.body.pos match {
-              case Position.None => c.body.pos.end
-              case Position.Range(input, _, end) =>
-                val char = input.chars(end)
-                if (char == '\r') end + 2
-                else end + 1
-            }
-          }
           for {
-            token <- startingPoint.findFirstTrailing(_.is[Token.RightArrow])
-            pos = range(tree.pos.input, token.pos.end, bodyEnd)
-          } yield (pos, true)
+            (foldStart, adjust) <- matchBlockStartAndFoldingAdjustment[
+              Token.RightArrow
+            ](
+              startingPoint.trailingTokens
+            )
+            pos = range(tree.pos.input, foldStart, c.body.pos.end)
+          } yield (pos, adjust)
 
         case term: Term.Try => // range for the `catch` clause
-          for {
-            startToken <- term.expr.findFirstTrailing(_.is[Token.KwCatch])
-            lastCase <- term.catchp.lastOption
-            endToken <- lastCase.findFirstTrailing(_.is[Token.RightBrace])
-            tryPos =
-              Position
-                .Range(tree.pos.input, startToken.pos.end, endToken.pos.end)
-          } yield (tryPos, true)
+          matchBlockStartAndFoldingAdjustment[Token.KwCatch](
+            term.expr.trailingTokens
+          ) match {
+            case Some(foldStartIndexAndAdjust) =>
+              Some(
+                range(
+                  tree.pos.input,
+                  foldStartIndexAndAdjust._1,
+                  term.pos.end,
+                ),
+                foldStartIndexAndAdjust._2,
+              )
+            case None => None
+          }
 
         case For(endPosition) =>
-          val start = tree.pos.start + 3
-          val end = endPosition.start
-          Some(range(tree.pos.input, start, end), true)
+          matchBlockStartAndFoldingAdjustment[Token.KwFor](
+            tree.tokens.iterator
+          ) match {
+            case Some(foldStartIndexAndAdjust) =>
+              Some(
+                range(
+                  tree.pos.input,
+                  foldStartIndexAndAdjust._1,
+                  endPosition.start,
+                ),
+                true,
+              )
+            case None => None
+          }
 
         case template: Template =>
+          // TODO is this case needed now if we handle folding from defn case?
```

Comment:
Is handling of `Template`s still needed here? Do they ever appear without `Defn` in front of them?

Reply:
Okay but `Class`, `Object` and `Trait` are handled in `case defn: Defn` so the case for `Template` won't be needed I think